storm-client/src/jvm/org/apache/storm/task/DefaultTaskNetworkDistanceCalculator.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.task;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.storm.Config;
+import org.apache.storm.cluster.IStormClusterState;
+import org.apache.storm.generated.NodeInfo;
+import org.apache.storm.networktopography.DNSToSwitchMapping;
+import org.apache.storm.utils.ReflectionUtils;
+
+/**
+ * This calculator calculates the network distance between tasks based on their physical locations:
+ * WORKER_LOCAL, HOST_LOCAL, RACK_LOCAL, UNKNOWN.
+ */
+public class DefaultTaskNetworkDistanceCalculator implements ITaskNetworkDistanceCalculator {
+
+    private DNSToSwitchMapping dnsToSwitchMapping;
+
+    public void prepare(Map<String, Object> conf, IStormClusterState stormClusterState) {
+        dnsToSwitchMapping = ReflectionUtils.newInstance((String) conf.get(Config.STORM_NETWORK_TOPOGRAPHY_PLUGIN));
+    }
+
+    public void calculateOrUpdate(Map<Integer, NodeInfo> taskToNodePort,
+                                  ConcurrentHashMap<Integer, ConcurrentHashMap<Integer, Double>> taskNetworkDistance) {
+
+        Map<String, String> hostToRack = getHostToRackMapping(taskToNodePort);
+
+        for (Map.Entry<Integer, NodeInfo> entry: taskToNodePort.entrySet()) {
+            Integer sourceTaskId = entry.getKey();
+            NodeInfo sourceNodeInfo = entry.getValue();
+
+            ConcurrentHashMap<Integer, Double> targetTaskDistances;
+            if (!taskNetworkDistance.containsKey(sourceTaskId)) {
+                targetTaskDistances = new ConcurrentHashMap<>();
+                taskNetworkDistance.put(sourceTaskId, targetTaskDistances);
+            } else {
+                targetTaskDistances = taskNetworkDistance.get(sourceTaskId);
+                targetTaskDistances.clear();
+            }
+
+            for (Map.Entry<Integer, NodeInfo> targetEntry: taskToNodePort.entrySet()) {
+                Integer targetTaskId = targetEntry.getKey();
+                NodeInfo targetNodeInfo = targetEntry.getValue();
+
+                // storm task doesn't transmit tuples to itself
+                if (!sourceTaskId.equals(targetTaskId)) {
+                    targetTaskDistances.put(targetTaskId,
+                            calculateDistance(sourceNodeInfo, targetNodeInfo, hostToRack).distance());
+                }
+            }
+
+        }
+    }
+
+    private DistanceType calculateDistance(NodeInfo source, NodeInfo target, Map<String, String> hostToRack) {
+        if(source.get_port().equals(target.get_port())) {
```

Comment:
@govind-menon  Thanks. I addressed your review. Could you please review the latest commit?